### PR TITLE
feat(compiler): dep-key callbacks for hooks that declare them as data

### PR DESCRIPTION
**Pull request by trueadm**
## Summary

New compile option `dataCallbackHooks`, letting a hook declare that a callback argument is **data it memoizes on** rather than a dependency subject it re-runs. Production compiles then key that callback on the values it actually reads.

This is option 1 from the earlier discussion, and it unparks the dep-keying work that was blocked on exactly this question.

## Why

The capture-free lift (#337) handles callbacks that capture nothing. A callback that reads component state cannot move to module scope — but its identity still only needs to change when the values it reads change.

Measured, 512 subscribers over 20 unrelated parent re-renders with a capturing selector:

| | selector invocations | prod re-render |
| --- | --- | --- |
| authored form | 10,240 | 21.4–23.1 ms |
| dep-keyed | **0** | 7.8–9.9 ms |

**The pass runs before dependency inference, and that ordering is the whole design.** Coarse identifier dependencies (`[props]`) are worthless here — I measured it, and the win vanishes completely (10,240 invocations, 24.2ms, identical to the authored form), because the props object is a fresh identity on every parent render so the memo never hits. Inference produces the member paths actually read (`[props.offset]`), which is what makes the identity hold.

## Why it is declared and not inferred

The fact lives in the hook's own package, and a single-module compile cannot read it. The cross-module facts spike (#321) that would have supplied it is closed.

More importantly, guessing is the wrong trade. The compiler already refuses to attach dependency semantics to custom hooks it cannot statically prove (`infers only statically proven local custom dependency hooks`), and a wrong answer here is a **stale closure** — silent, intermittent, and attributed to the application rather than to the compiler — where the lift's failure mode is a loud `ReferenceError`. So the transform is **inert until something opts in**.

```js
compile(source, filename, {
  dataCallbackHooks: ['@octanejs/tanstack-store#useSelector'],
})
```

Entries are `module#hookName`, matched against the call site's import, or a bare `hookName` for a hook declared in the module being compiled.

Declared hooks are still skipped wherever the hook owns freshness itself: when its own dependency list is inferred (wrapping there turns a working inference into a hard compile error — found by the suite), and when the call passes an explicit dependency array or `null`, the author's "re-run every render" escape hatch. Dev, HMR and profile compiles keep the authored form.

## Validation

- [x] `pnpm format:check`
- [x] `pnpm typecheck`
- [x] `pnpm test` — **1438 files, 14,641 tests, 0 failures**

Coverage is split deliberately, because the vitest projects cannot pass a per-file compile option:

- **Codegen** (12 assertions) — inert by default; wraps once declared; the dependency list is the member path `[props.id]`; module-scoped matching rejects a same-named hook from another package; bare names work for local hooks; explicit `null`/array deps are skipped; dev, HMR and profile are skipped.
- **Behavioral** (4, across both compile modes) — the *emitted shape* is exercised directly, hand-written to match what the compiler produces, against the authored form as a control. It asserts the selection tracks the store, the captured prop, and unrelated local state independently. This is what proves dep-keying does not go stale; the codegen tests prove the compiler emits that shape.

## Risk / follow-ups

- **Not yet exposed through the Vite/Rspack plugins.** The option is compile-level and tested there; plugin passthrough is the remaining step before an app can actually opt in, and I did not want to guess at that wiring. No binding declares itself yet, so nothing changes for anyone today.
- The lift's safety analysis is now property-tested (#342), which was the prerequisite for trusting this transform at all.
- Uses `useCallback` at the argument position rather than a preceding `const`, so the inline hook-memo tier does not consume it and it survives as a runtime call. Measured at parity with the declaration form (9.4 vs 9.9ms), and it keeps the rewrite local — no statement insertion, so no risk of lifting a callback out of the block whose scope it depends on.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Production-only AST rewrite can introduce stale closures if `dataCallbackHooks` is mis-declared, but the transform is opt-in and skips explicit author dependency escape hatches and inference-owned hook calls.
> 
> **Overview**
> Adds an opt-in compile option **`dataCallbackHooks`** so production builds can stabilize callback identities for hooks that memoize on their callback argument (e.g. store selectors), complementing the existing capture-free module-scope lift.
> 
> For declared hooks only, capturing inline callbacks are wrapped inline as **`_$autoCb(...)`** (aliased `useCallback` from `octane`) **before** hook dependency inference runs, so inferred deps use member paths like **`[props.id]`** instead of useless coarse **`[props]`**. Capture-free callbacks still hoist to module scope; calls with explicit dependency arrays, `null`, or TypeScript-wrapped equivalents are left alone, as are hooks whose own deps are inference-owned.
> 
> Matching is explicit: **`module#hookName`**, namespace member calls against the namespace module, or a bare name for a hook defined in the compiling module—default/namespace imports from other packages never match a bare entry. The transform is disabled for dev, HMR, and profile compiles, matching the inline hook-memo tier.
> 
> Codegen and behavioral tests cover inert-by-default behavior, wrapping shape, import matching edge cases, and semantic parity between authored and dep-keyed selector usage.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 4726686117284af35f6f75282cddf378e246af69. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/octane/src/compiler/compile.js [hunk 1]
```diff
@@ -3056,6 +3066,169 @@ function hoistCaptureFreeHookCallbacks(ast, options = {}) {
...
+	/** Whether THIS call site's hook declared its callback is data. */
+	const hookDeclaresDataCallback = (call) => {
+		const name = hookShapedCalleeName(call);
+		if (name === null) return false;
+		const origin = importOrigins.get(name);
+		if (origin !== undefined) {
+			return declared.byModule.get(origin.module)?.has(origin.imported) === true;
+		}
+		// Not an import: a hook declared in this module.
+		return declared.bareNames.has(name);
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js]: Confirmed and fixed in c08ab06f. Reproduced first — with a bare `useSelector` declaration, both of these were wrapped despite coming from a package that never opted in:

```js
import useSelector from 'some-other-package';          // wraps=1
import * as Store from 'some-other-package';           // Store.useSelector(...) → wraps=1
```

Root cause is as described. `collectImportOrigins` recorded only `ImportSpecifier`, so a default or namespace local was absent from the map, and the predicate read "absent from imports" as "declared locally". `hookShapedCalleeName` collapsing a member call to its property name is what let the namespace case through.

Fixed by attributing each callee to how it actually entered the module:

- import origins now record named, namespace, and default forms separately;
- a namespace member call is matched against **that namespace's own module**, so `Store.useSelector` opts in under `@octanejs/tanstack-store#useSelector` and not under a bare entry;
- a bare entry matches only a plain identifier that is not imported at all;
- any other callee shape is not declarable.

This mattered more than the Medium label suggests: the whole point of requiring a declaration is that a wrong answer here is a stale closure rather than a loud error, and this let a hook that never declared anything be dep-keyed.

Two regression tests added — default and namespace imports from another package must not match a bare entry, and a namespace import must still match its own module. Both verified failing against the pre-fix predicate.

File: packages/octane/src/compiler/compile.js [hunk 2]
```diff
@@ -3056,6 +3066,169 @@ function hoistCaptureFreeHookCallbacks(ast, options = {}) {
...
+				if (
+					next &&
+					(next.type === 'ArrayExpression' ||
+						(next.type === 'Literal' && next.value === null) ||
+						next.type === 'NullLiteral' ||
+						// `undefined` parses as an Identifier, and passing it explicitly
+						// means the same as omitting the list.
+						(next.type === 'Identifier' && next.name === 'undefined'))
+				) {
+					return walked;
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js:3241]: Agreed, fixed in c08ab06f — the code and the comment beside it genuinely disagreed, and the comment was right.

`undefined` was originally added to that skip set while the transform was ungated, to satisfy a dependency-inference codegen test. Once the `inferenceOwned` check landed, that case was already covered by it, and the `undefined` clause was left behind doing harm: `useSelector(store, selector, undefined)` is a realistic shape for a hook whose third parameter is options, and it silently kept the authored identity churn.

Only an explicit array or `null` now counts as an author-owned dependency list. Those are deliberate; `undefined` is what an omitted optional argument looks like.

Regression test added for `useSelector(store, fn, undefined)`, verified failing against the previous skip set. The `null` and explicit-array cases still skip, and the inference-owned guard is unchanged.

File: packages/octane/src/compiler/compile.js [hunk 3]
```diff
@@ -3056,6 +3066,197 @@ function hoistCaptureFreeHookCallbacks(ast, options = {}) {
...
+				// render" escape hatch, so memoizing the callback beneath it could
+				// hand the hook a stale closure.
+				const next = n.arguments[index + 1];
+				if (
+					next &&
+					(next.type === 'ArrayExpression' ||
+						(next.type === 'Literal' && next.value === null) ||
+						next.type === 'NullLiteral')
+				) {
+					return walked;
```
[the remark this answers is not in the document]
Comment (trueadm) [reply packages/octane/src/compiler/compile.js:3241]: Confirmed and fixed in 47266861. Reproduced first — with the hook declared, three of these bypassed the escape hatch:

| deps written as | before | after |
| --- | --- | --- |
| `null` | skipped | skipped |
| `(null)` | skipped | skipped |
| `null as any` | **wrapped** | skipped |
| `[props.id] as const` | **wrapped** | skipped |
| `[props.id]!` | **wrapped** | skipped |

Peeled with the existing module-level `unwrapTsExpr` rather than adding a fourth copy of that node list — thanks for pointing at the precedent, it already covers `TSAsExpression`, `TSNonNullExpression`, `TSTypeAssertion`, `TSSatisfiesExpression`, and `ParenthesizedExpression`.

Regression test covers all four wrapper forms and is verified failing against the unpeeled check.

Worth recording that this is the third finding in a row on this PR with the same shape: the declaration gate reading a *syntactic* form and missing the ways that form can be written. Import origins missed default and namespace imports; `undefined` was read as a deliberate list; and now type wrappers hid the list entirely. Each one fails toward wrapping — the silent stale-closure direction rather than the loud one — which is an argument for the gate being conservative by construction rather than by enumeration. I have not restructured it here, since the enumeration is now complete as far as I can probe it, but it is the thing I would look at first if a fourth turns up.
